### PR TITLE
SWC binding for Jest and React update

### PR DESCRIPTION
**Pull request by nikomakela**
Updated the React, Typescript, Eslint and the Jest runner! All this because I wanted to start using the SWC, but it required updates which required more updates...

Installed a SWC binding for Jest and a plugin to handle Jest compatibility issues. This approach was already tested in the events-helsinki-monorepo: City-of-Helsinki/events-helsinki-monorepo#197. The tests are now twice as fast as before (but some of them are still quite slow).

### CODE REVIEW

**Comment by nikomakela** [comment]
Some build issues still: https://github.com/City-of-Helsinki/palvelutarjotin-ui/actions/runs/4305562013/jobs/7508179359. Didn't have any of these in my local environment, so there might be something related to the docker image?

- [ ] Fix the build time issues

**Comment by terovirtanen** [comment]
**kultus-ui branch is deployed to platta: https://kultus-ui-pr269.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**kultus-ui branch is deployed to platta: https://kultus-ui-pr269.dev.hel.ninja** :rocket::rocket::rocket:

**Review by karisal-anders — COMMENTED** [review]
Did you mean to change from node 14 to node 16 everywhere? Searching with `node.*14` found references in two files:
 - README.md → `- Node 14.x`
 - staging.yml → `node-version: '14'`

**Review by karisal-anders — COMMENTED** [review]
Any ideas what could be done to make the test `renders search form and events list with correct information` more robust? Running `yarn test` with all tests consistently fails on my machine—tried 5 times and the running time for the failing test `renders search form and events list with correct information` varied between ~88s–98s:
```typescript
Summary of all failing tests
 FAIL  src/domain/events/__tests__/EventsPage.test.tsx (98.414 s)
  ● renders search form and events list with correct information

    thrown: "Exceeded timeout of 70000 ms for a testfalse.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test."

      138 | });
      139 |
    > 140 | test('renders search form and events list with correct information', async () => {
          | ^
      141 |   advanceTo(testDate);
      142 |   render(<EventsPage />, { mocks });
      143 |

      at Object.test (src/domain/events/__tests__/EventsPage.test.tsx:140:1)


Test Suites: 1 failed, 1 skipped, 50 passed, 51 of 52 total
Tests:       1 failed, 3 skipped, 209 passed, 213 total
Snapshots:   46 passed, 46 total
Time:        150.758 s, estimated 189 s
Ran all test suites.
```

**Comment by nikomakela** [comment]
>  Running yarn test with all tests consistently fails on my machine—tried 5 times and the running time for the failing test renders search form and events list with correct information varied between ~88s–98s

Are you sure you have installed the new packages (with `yarn`)? I'm asking because on my machine, all the 52 test suites are run in 100s:

```
Test Suites: 1 skipped, 51 passed, 51 of 52 total
Tests:       3 skipped, 210 passed, 213 total
Snapshots:   46 passed, 46 total
Time:        89.514 s
Ran all test suites.
```

```
Test Suites: 1 skipped, 51 passed, 51 of 52 total
Tests:       3 skipped, 210 passed, 213 total
Snapshots:   46 passed, 46 total
Time:        90.353 s
Ran all test suites.
```

If I run that EventsPage.test that you raised up, it's run in 30s. Yea, it's still much, but it's a huge test. It's been there for a while now and I remember we tried to make it smaller with Santeri earlier, but didn't find a way yet. Sure the suite could be splitted, but that would mean much more code and harder to read tests. That is something to consider.

```
Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        28.637 s, estimated 28 s
Ran all test suites matching /eventspage/i.
```

**Comment by terovirtanen** [comment]
**kultus-ui branch is deployed to platta: https://kultus-ui-pr269.dev.hel.ninja** :rocket::rocket::rocket:

File: .eslintrc.json [hunk 1]
```diff
@@ -27,7 +26,7 @@
     "@typescript-eslint/func-call-spacing": ["error"],
     "@typescript-eslint/member-ordering": ["warn"],
     "@typescript-eslint/no-require-imports": ["error"],
-    "react/no-unused-prop-types": ["warn", { "skipShapeProps": true }],
+    "react/no-unused-prop-types": "off",
```
Comment (karisal-anders) [inline .eslintrc.json:29]: Why was this turned off?

File: src/domain/event/occurrences/OccurrencesTable.tsx [hunk 2]
```diff
@@ -314,6 +314,7 @@ const OccurrenceEnrolmentTable: React.FC<{
 
   return (
     <Table
+      // @ts-ignore
```
Comment (karisal-anders) [inline src/domain/event/occurrences/OccurrencesTable.tsx:317]: Why was this added? [Typescript documentation](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html#suppress-errors-in-ts-files-using--ts-ignore-comments) says:
> It is recommended practice to have the remainder of the comment following @ts-ignore explain which error is being suppressed.

File: src/domain/matomo/Matomo.tsx [hunk 3]
```diff
@@ -25,6 +25,7 @@ const matomoInstance = createMatomoInstance({
 function Matomo({ children }: { children: React.ReactNode }): JSX.Element {
   if (matomoInstance) {
     return (
+      // @ts-ignore 2322
```
Comment (karisal-anders) [inline src/domain/matomo/Matomo.tsx:28]: This is not needed if matomo-tracker-react is changed from @datapunt/matomo-tracker-react to @jonkoops/matomo-tracker-react

File: src/domain/event/occurrences/OccurrencesTable.tsx (hunk 2 above)
Comment (karisal-anders) [reply src/domain/event/occurrences/OccurrencesTable.tsx:317]: Adding `"noErrorTruncation": true` to `compilerOptions` in tsconfig.json makes it possible to see the entire error message without @ts-ignore:

`> npx tsc --project ./tsconfig.json --noEmit`:

```typescript
src/domain/event/occurrences/OccurrencesTable.tsx:317:7 - error TS2322: Type '({ id: string; Header: string; 'aria-hidden': boolean; accessor?: undefined; Cell?: undefined; style?: undefined; } | { accessor: (row: OccurrenceFieldsFragment) => string | null; id: string; Header?: undefined; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { Header: string; accessor: (row: OccurrenceFieldsFragment) => Element | "–"; id: string; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { Header: string; accessor: (row: OccurrenceFieldsFragment) => Element | "-"; id: string; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { Header: string; accessor: (row: OccurrenceFieldsFragment) => string; id: string; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { accessor: (row: OccurrenceFieldsFragment) => OccurrenceFieldsFragment; Cell: ({ row, value, }: { value: OccurrenceFieldsFragment; row: { getToggleRowExpandedProps: () => ButtonProps & RefAttributes<HTMLButtonElement>; isExpanded: boolean; }; }) => Element; id: string; style: { width: string; }; Header?: undefined; 'aria-hidden'?: undefined; })[]' is not assignable to type 'Column<OccurrenceFieldsFragment>[]'.
  Type '{ id: string; Header: string; 'aria-hidden': boolean; accessor?: undefined; Cell?: undefined; style?: undefined; } | { accessor: (row: OccurrenceFieldsFragment) => string | null; id: string; Header?: undefined; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { Header: string; accessor: (row: OccurrenceFieldsFragment) => Element | "–"; id: string; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { Header: string; accessor: (row: OccurrenceFieldsFragment) => Element | "-"; id: string; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { Header: string; accessor: (row: OccurrenceFieldsFragment) => string; id: string; 'aria-hidden'?: undefined; Cell?: undefined; style?: undefined; } | { accessor: (row: OccurrenceFieldsFragment) => OccurrenceFieldsFragment; Cell: ({ row, value, }: { value: OccurrenceFieldsFragment; row: { getToggleRowExpandedProps: () => ButtonProps & RefAttributes<HTMLButtonElement>; isExpanded: boolean; }; }) => Element; id: string; style: { width: string; }; Header?: undefined; 'aria-hidden'?: undefined; }' is not assignable to type 'Column<OccurrenceFieldsFragment>'.
    Type '{ accessor: (row: OccurrenceFieldsFragment) => OccurrenceFieldsFragment; Cell: ({ row, value, }: { value: OccurrenceFieldsFragment; row: {    getToggleRowExpandedProps: () => React.ComponentProps<typeof Button>;    isExpanded: boolean;}; }) => JSX.Element; id: string; style: { width: string; }; Header?: undefined; 'aria-hidden'?: undefined; }' is not assignable to type 'Column<OccurrenceFieldsFragment>'.
      Type '{ accessor: (row: OccurrenceFieldsFragment) => OccurrenceFieldsFragment; Cell: ({ row, value, }: { value: OccurrenceFieldsFragment; row: {    getToggleRowExpandedProps: () => React.ComponentProps<typeof Button>;    isExpanded: boolean;}; }) => JSX.Element; id: string; style: { width: string; }; Header?: undefined; 'aria-hidden'?: undefined; }' is not assignable to type 'ColumnInterface<OccurrenceFieldsFragment> & { accessor: "seatsTaken"; } & ColumnInterfaceBasedOnValue<OccurrenceFieldsFragment, number>'.
        Type '{ accessor: (row: OccurrenceFieldsFragment) => OccurrenceFieldsFragment; Cell: ({ row, value, }: { value: OccurrenceFieldsFragment; row: {    getToggleRowExpandedProps: () => React.ComponentProps<typeof Button>;    isExpanded: boolean;}; }) => JSX.Element; id: string; style: { width: string; }; Header?: undefined; 'aria-hidden'?: undefined; }' is not assignable to type '{ accessor: "seatsTaken"; }'.
          Types of property 'accessor' are incompatible.
            Type '(row: OccurrenceFieldsFragment) => OccurrenceFieldsFragment' is not assignable to type '"seatsTaken"'.

317       columns={columns}
          ~~~~~~~

  src/common/components/table/Table.tsx:9:3
    9   columns: Array<Column<D>>;
        ~~~~~~~
    The expected type comes from property 'columns' which is declared here on type 'IntrinsicAttributes & Props<OccurrenceFieldsFragment>'


Found 1 error in src/domain/event/occurrences/OccurrencesTable.tsx:317
```

Comment (nikomakela) [reply src/domain/event/occurrences/OccurrencesTable.tsx:317]: I investigated it for a while earlier. It might be that the 3rd party TableComponent that we used there does not support React 18 properly. The feature itself seems to work, so I just added the `@ts-ignore` and didn't want to waste more time with that at this point. But if you are eager to find a fix for that and find one, that's a good thing! :)

File: .eslintrc.json (hunk 1 above)
Comment (nikomakela) [reply .eslintrc.json:29]: there were a lot of those errors in some 3rd party plugin usages if I remember right. My idea was to make a minimum upgrade to get the SCW-Jest to work. I think this could have been fixed by upgrading rest of the packages as well.

File: src/domain/matomo/Matomo.tsx (hunk 3 above)
Comment (karisal-anders) [reply src/domain/matomo/Matomo.tsx:28]: IMHO yes, it's an easy fix, nothing else but `yarn remove @datapunt/matomo-tracker-react` and `yarn add @jonkoops/matomo-tracker-react` and replacing all `@datapunt` with `@jonkoops` in the source code and it should be done. BTW @jonkoops/matomo-tracker-react is the one used in events-helsinki-monorepo.

About the packages one possible problem is that [@jonkoops/matomo-tracker-react](https://github.com/jonkoops/matomo-tracker) v0.7.0 is the last release from 2022-05-23 and the repository is no longer maintained. 

snyk.io gives health score [53/100](https://snyk.io/advisor/npm-package/@datapunt/matomo-tracker-react) for @datapunt/matomo-tracker-react v0.5.1 and [63/100](https://snyk.io/advisor/npm-package/@jonkoops/matomo-tracker-react) for @jonkoops/matomo-tracker-react v0.7.0 so that is slightly in favor of the latter.

@datapunt/matomo-tracker-react v0.5.1 was last published 2021-11-08 according to [npmjs.com](https://www.npmjs.com/package/@datapunt/matomo-tracker-react).

Taking all of these into account I would say changing to @jonkoops/matomo-tracker-react is reasonable even though it is not maintained anymore. I have not looked into any third alternative to these but if there would be that could be an option as well. But as this should be just a direct replacement and fixes the type issue here I think this is a reasonable thing to do.

File: src/domain/event/occurrences/OccurrencesTable.tsx (hunk 2 above)
Comment (karisal-anders) [reply src/domain/event/occurrences/OccurrencesTable.tsx:317]: @nikomakela Hopefully "yarn typecheck" passes for you locally as well now, I'm not sure what happened, the typescript error above disappeared at some point and didn't reappear, weird! Hopefully the problem is gone now!

**Comment by terovirtanen** [comment]
**kultus-ui branch is deployed to platta: https://kultus-ui-pr269.dev.hel.ninja** :rocket::rocket::rocket:

File: .eslintrc.json (hunk 1 above)
Comment (karisal-anders) [reply .eslintrc.json:29]: If I added the `"react/no-unused-prop-types": ["warn", { "skipShapeProps": true }],` back now I got:
```typescript
palvelutarjotin-ui> yarn lint
yarn run v1.22.19
$ eslint --ext js,ts,tsx src

palvelutarjotin-ui\src\common\components\datepicker\Datepicker.tsx
  38:3  warning  'disabled' PropType is defined but prop is never used  react/no-unused-prop-types

palvelutarjotin-ui\src\common\components\datepicker\TimesList.tsx
  11:3  warning  'times' PropType is defined but prop is never used        react/no-unused-prop-types
  12:3  warning  'onTimeClick' PropType is defined but prop is never used  react/no-unused-prop-types
  13:3  warning  'datetime' PropType is defined but prop is never used     react/no-unused-prop-types

palvelutarjotin-ui\src\domain\event\occurrences\EnrolmentButtonCell.tsx
  21:3  warning  'event' PropType is defined but prop is never used                react/no-unused-prop-types
  22:3  warning  'value' PropType is defined but prop is never used                react/no-unused-prop-types
  23:3  warning  'selectedOccurrences' PropType is defined but prop is never used  react/no-unused-prop-types
  24:3  warning  'enrolOccurrence' PropType is defined but prop is never used      react/no-unused-prop-types
  25:3  warning  'selectOccurrence' PropType is defined but prop is never used     react/no-unused-prop-types
  26:3  warning  'deselectOccurrence' PropType is defined but prop is never used   react/no-unused-prop-types
  27:3  warning  'neededOccurrences' PropType is defined but prop is never used    react/no-unused-prop-types

✖ 11 problems (0 errors, 11 warnings)
```

Thinking whether those could easily be tackled in some other way...

File: src/domain/event/occurrences/OccurrencesTable.tsx (hunk 2 above)
Comment (karisal-anders) [reply src/domain/event/occurrences/OccurrencesTable.tsx:317]: Tried cloning the palvelutarjotin-ui repository afresh, ran "yarn install" and "yarn typecheck" and the typecheck went through without any warnings/errors so it should be fine I think.

File: .eslintrc.json (hunk 1 above)
Comment (karisal-anders) [reply .eslintrc.json:29]: - Datepicker doesn't support `disabled` at all so that property could be removed
- TimesList itself does use all of those props (times, onTimeClick and datetime), that might be [false positive from ESLint](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unused-prop-types.md#known-issueslimitations). Let's see if I can get this to work somehow...
- EnrolmentButtonCell.tsx is unused so that file could be removed

**Review by karisal-anders — APPROVED** [review]

File: .eslintrc.json (hunk 1 above)
Comment (nikomakela) [reply .eslintrc.json:29]: Could be fixed yea! Since my winter holiday is starting, I would like to merge this now. Lets fix these in another PR.

**Comment by terovirtanen** [comment]
**kultus-ui branch is deployed to platta: https://kultus-ui-pr269.dev.hel.ninja** :rocket::rocket::rocket:
